Multiple database support

wesleykendall opened the pull request:
Supports multiple-database functionality in all core functions and management commands.
By default, all functions and management commands operate over all databases in a
multi-database setup. This behavior can be overridden with the ``--database`` flag.

When calling ``manage.py migrate``, only the database being migrated will have
relevant triggers installed. This fits into how Django supports multi-database
migrations.

Type: feature

## Overview

The issue #8 opened up the possibility of this library working with multiple databases. After playing around with some solutions, I eventually settled on using the database router config as the source of truth for where triggers for a particular model are installed.

Django database routers allow one to specify the default databased used for writing models. It also allows one to configure where models are migrated. Some comments in #8 suggested using router migration configuration to determine where triggers are installed, but this opened up a can of worms since models can be migrated on multiple databases. I chose using the configured write DB as the DB where triggers are installed for a particular model. This means triggers for a model can only be installed on one database.

I'm open to hear if this will be limiting for certain situations, but it is difficult for me to think of a scenario where one is writing changes to the same model across two different databases. I'm not entirely sure if you could accomplish this with Django.

Using this strategy allowed me to keep this change pretty clean. There is one utility function for determining the database for a model, and all trigger functionality looks up the database when applying a trigger operation for a particular model.

The management commands (and core interface) work across all databases by default. For example, `pgtrigger install` will install triggers across all configured databases. One can use `pgtrigger install --database name` in order to install triggers for a particular database.

`pgtrigger ls` was updated to also list the database of the trigger in the second column of the output. It also supports the ``--database`` option for targeting one database.

By default, triggers are installed after ``manage.py migrate`` unless one ignores this functionality with a configured setting. Since ``manage.py migrate`` only migrates one database at a time (i.e. it uses the "default" database when no ``--database`` argument is supplied), the migration hook in pgtrigger will only install triggers for the particular database migrated.

## TODO

There are still quite a few docs to update. I'm also not sure of the best way to test this. I'm going to add a few more databases in the testing configuration so I can be certain this is working as intended.

I will likely not be able to come back to this PR for a few days. It would be nice if people from #8 could try to check out this code and see if it works for their use case. Happy to provide more explanation

wesleykendall commented:
@tomage can you check this one out?

arpitremarkable commented on pgtrigger/core.py:
```diff
@@ -31,6 +34,31 @@
+    for the connection is selected based on the write DB in the database
+    router config.
+    """
+    return router.db_for_write(model) or 'default'
```
Should use `from django.db import DEFAULT_DB_ALIAS`

wesleykendall replied:
Good call! Will fix

tomage commented on pgtrigger/core.py:
```diff
@@ -800,12 +837,22 @@ def _render_value():
     A URI is in the format of "app_label.model_name:trigger_name"
     """
+    if database and uris:
+        raise ValueError('Cannot supply both trigger URIs and a database')
```
I've usually had somewhat of a hard time finding the right built-in exceptions to raise in various circumstances - I'm fine using ValueError here - but was curious if you thought `RuntimeError` would be applicable?

wesleykendall replied:
Yea I have had a difficult time too. And finding answers like these https://stackoverflow.com/questions/256222/which-exception-should-i-raise-on-bad-illegal-argument-combinations-in-python does not help add any clarity.

In that thread, there were a couple answers that suggested using a ValueError for invalid argument combinations. Others suggested TypeError and then others suggested a custom exception. I have just always used ValueError when the values in the arguments are invalid or when the combination of arguments are invalid. It's just a style thing for me.

I've always used RuntimeError as a catch all when something goes wrong in the middle of running a function

tomage replied:
That sounds totally fine to me! I read thru python's docs as well on exceptions, and there's no clear answer there.

tomage commented on pgtrigger/core.py:
```diff
@@ -820,140 +867,175 @@ def get(*uris):
+    if isinstance(database, str):
+        databases = [database]
+    else:
+        databases = database or settings.DATABASES
```
Say - could `database` not be a singular object here?
I.e. should this be:
```
elif database is None:
    databases = settings.DATABASES
```

I perhaps haven't fully grokked the possible values for `database`.

wesleykendall replied:
I'm only handling the case of:

1. A single string (e.g. "default")
2. An iterable of strings

When a false-y object is provided, it defaults to all databases (i.e the dictionary keys of the DATABASES setting)

None of the management commands allow providing multiple databases (yet), but I went ahead and just made all of these functions work with singular or multiple databases if we extend the management commands to take multiple ``--database`` arguments.

If you passed in a random object or iterable of non-strings, it would just fail unexpectedly.

tomage replied:
Ah doi - the first `if isinstance(database, str):` takes care of the singular case anyway 🤦

tomage reviewed (commented):
Couple of comments - but otherwise I think this is more or less solid!

Thanks for a great explanation on decisions made in the description, def helps parsing the PR.

wesleykendall commented:
Addressed comments

@tomage I also added some basic docs for multi database support. I added a note about support for sharded instances later and basically instructed people to open an issue if this is their database setup.

@tomage I also add a basic test suite for multiple database support. pytest-django does not support multiple databases, so I had to go back to a ``TestCase`` setup. These tests do a few basic setups with the core management commands to verify installation and other core commands.

This should be good to go! The commit history has been formatted to bump this up a major version. There were no breaking changes to the best of my knowledge.

tomage reviewed (approved):
Yep! Looks clean - even with the bits to support multi-DB tests - must have been a fun challenge to get past those  :)
